### ISSUE TITLE
Tenant aware jobs from redis queue querying wrong tenants database

### ISSUE THREAD

**Issue by slavatarsyurimi** [spatie/laravel-multitenancy#426]
Hello,

Recently we've been recieving a very large number of `Illuminate\Database\QueryException` exceptions with the message:
`SQLSTATE[42S02]: Base table or view not found: 1146 Table 'application_x_tenant_2.tenants' doesn't exist (SQL: select * from 'tenants' where 'tenants'.'id' = 2 limit 1)`

The environment: 
- php **8.1.5**
- Laravel **v9.44.0**
- queue connection: **redis**
- package version: **2.3.9**
- `'queues_are_tenant_aware_by_default' => true`

Full stack trace:
```
PDOException: SQLSTATE[42S02]: Base table or view not found: 1146 Table 'application_x_tenant_2.tenants' doesn't exist
#39 /vendor/laravel/framework/src/Illuminate/Database/Connection.php(413): PDO::prepare
#38 /vendor/laravel/framework/src/Illuminate/Database/Connection.php(413): Illuminate\Database\Connection::Illuminate\Database\{closure}
#37 /vendor/laravel/framework/src/Illuminate/Database/Connection.php(752): Illuminate\Database\Connection::runQueryCallback
#36 /vendor/laravel/framework/src/Illuminate/Database/Connection.php(719): Illuminate\Database\Connection::run
#35 /vendor/laravel/framework/src/Illuminate/Database/Connection.php(421): Illuminate\Database\Connection::select
#34 /vendor/laravel/framework/src/Illuminate/Database/Query/Builder.php(2630): Illuminate\Database\Query\Builder::runSelect
#33 /vendor/laravel/framework/src/Illuminate/Database/Query/Builder.php(2618): Illuminate\Database\Query\Builder::Illuminate\Database\Query\{closure}
#32 /vendor/laravel/framework/src/Illuminate/Database/Query/Builder.php(3154): Illuminate\Database\Query\Builder::onceWithColumns
#31 /vendor/laravel/framework/src/Illuminate/Database/Query/Builder.php(2619): Illuminate\Database\Query\Builder::get
#30 /vendor/laravel/framework/src/Illuminate/Database/Eloquent/Builder.php(698): Illuminate\Database\Eloquent\Builder::getModels
#29 /vendor/laravel/framework/src/Illuminate/Database/Eloquent/Builder.php(682): Illuminate\Database\Eloquent\Builder::get
#28 /vendor/laravel/framework/src/Illuminate/Database/Concerns/BuildsQueries.php(296): Illuminate\Database\Eloquent\Builder::first
#27 /vendor/laravel/framework/src/Illuminate/Database/Eloquent/Builder.php(430): Illuminate\Database\Eloquent\Builder::find
#26 /vendor/laravel/framework/src/Illuminate/Support/Traits/ForwardsCalls.php(23): Illuminate\Database\Eloquent\Model::forwardCallTo
#25 /vendor/laravel/framework/src/Illuminate/Database/Eloquent/Model.php(2176): Illuminate\Database\Eloquent\Model::__call
#24 /vendor/laravel/framework/src/Illuminate/Database/Eloquent/Model.php(2188): Illuminate\Database\Eloquent\Model::__callStatic
#23 /vendor/spatie/laravel-multitenancy/src/Actions/MakeQueueTenantAwareAction.php(111): Spatie\Multitenancy\Actions\MakeQueueTenantAwareAction::findTenant
#22 /vendor/spatie/laravel-multitenancy/src/Actions/MakeQueueTenantAwareAction.php(48): Spatie\Multitenancy\Actions\MakeQueueTenantAwareAction::Spatie\Multitenancy\Actions\{closure}
#21 /vendor/laravel/framework/src/Illuminate/Events/Dispatcher.php(421): Illuminate\Events\Dispatcher::Illuminate\Events\{closure}
#20 /vendor/laravel/framework/src/Illuminate/Events/Dispatcher.php(249): Illuminate\Events\Dispatcher::dispatch
#19 /vendor/laravel/framework/src/Illuminate/Queue/Worker.php(618): Illuminate\Queue\Worker::raiseBeforeJobEvent
#18 /vendor/laravel/framework/src/Illuminate/Queue/Worker.php(416): Illuminate\Queue\Worker::process
#17 /vendor/laravel/framework/src/Illuminate/Queue/Worker.php(379): Illuminate\Queue\Worker::runJob
#16 /vendor/laravel/framework/src/Illuminate/Queue/Worker.php(173): Illuminate\Queue\Worker::daemon
#15 /vendor/laravel/framework/src/Illuminate/Queue/Console/WorkCommand.php(150): Illuminate\Queue\Console\WorkCommand::runWorker
#14 /vendor/laravel/framework/src/Illuminate/Queue/Console/WorkCommand.php(134): Illuminate\Queue\Console\WorkCommand::handle
#13 /vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php(36): Illuminate\Container\BoundMethod::Illuminate\Container\{closure}
#12 /vendor/laravel/framework/src/Illuminate/Container/Util.php(41): Illuminate\Container\Util::unwrapIfClosure
#11 /vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php(93): Illuminate\Container\BoundMethod::callBoundMethod
#10 /vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php(37): Illuminate\Container\BoundMethod::call
#9 /vendor/laravel/framework/src/Illuminate/Container/Container.php(651): Illuminate\Container\Container::call
#8 /vendor/laravel/framework/src/Illuminate/Console/Command.php(139): Illuminate\Console\Command::execute
#7 /vendor/symfony/console/Command/Command.php(308): Symfony\Component\Console\Command\Command::run
#6 /vendor/laravel/framework/src/Illuminate/Console/Command.php(124): Illuminate\Console\Command::run
#5 /vendor/symfony/console/Application.php(998): Symfony\Component\Console\Application::doRunCommand
#4 /vendor/symfony/console/Application.php(299): Symfony\Component\Console\Application::doRun
#3 /vendor/symfony/console/Application.php(171): Symfony\Component\Console\Application::run
#2 /vendor/laravel/framework/src/Illuminate/Console/Application.php(102): Illuminate\Console\Application::run
#1 /vendor/laravel/framework/src/Illuminate/Foundation/Console/Kernel.php(129): Illuminate\Foundation\Console\Kernel::handle
#0 /artisan(37): null
Illuminate\Database\QueryException: SQLSTATE[42S02]: Base table or view not found: 1146 Table 'application_x_tenant_2.tenants' doesn't exist (SQL: select * from `tenants` where `tenants`.`id` = 1 limit 1)
#37 /vendor/laravel/framework/src/Illuminate/Database/Connection.php(759): Illuminate\Database\Connection::runQueryCallback
#36 /vendor/laravel/framework/src/Illuminate/Database/Connection.php(719): Illuminate\Database\Connection::run
#35 /vendor/laravel/framework/src/Illuminate/Database/Connection.php(421): Illuminate\Database\Connection::select
#34 /vendor/laravel/framework/src/Illuminate/Database/Query/Builder.php(2630): Illuminate\Database\Query\Builder::runSelect
#33 /vendor/laravel/framework/src/Illuminate/Database/Query/Builder.php(2618): Illuminate\Database\Query\Builder::Illuminate\Database\Query\{closure}
#32 /vendor/laravel/framework/src/Illuminate/Database/Query/Builder.php(3154): Illuminate\Database\Query\Builder::onceWithColumns
#31 /vendor/laravel/framework/src/Illuminate/Database/Query/Builder.php(2619): Illuminate\Database\Query\Builder::get
#30 /vendor/laravel/framework/src/Illuminate/Database/Eloquent/Builder.php(698): Illuminate\Database\Eloquent\Builder::getModels
#29 /vendor/laravel/framework/src/Illuminate/Database/Eloquent/Builder.php(682): Illuminate\Database\Eloquent\Builder::get
#28 /vendor/laravel/framework/src/Illuminate/Database/Concerns/BuildsQueries.php(296): Illuminate\Database\Eloquent\Builder::first
#27 /vendor/laravel/framework/src/Illuminate/Database/Eloquent/Builder.php(430): Illuminate\Database\Eloquent\Builder::find
#26 /vendor/laravel/framework/src/Illuminate/Support/Traits/ForwardsCalls.php(23): Illuminate\Database\Eloquent\Model::forwardCallTo
#25 /vendor/laravel/framework/src/Illuminate/Database/Eloquent/Model.php(2176): Illuminate\Database\Eloquent\Model::__call
#24 /vendor/laravel/framework/src/Illuminate/Database/Eloquent/Model.php(2188): Illuminate\Database\Eloquent\Model::__callStatic
#23 /vendor/spatie/laravel-multitenancy/src/Actions/MakeQueueTenantAwareAction.php(111): Spatie\Multitenancy\Actions\MakeQueueTenantAwareAction::findTenant
#22 /vendor/spatie/laravel-multitenancy/src/Actions/MakeQueueTenantAwareAction.php(48): Spatie\Multitenancy\Actions\MakeQueueTenantAwareAction::Spatie\Multitenancy\Actions\{closure}
#21 /vendor/laravel/framework/src/Illuminate/Events/Dispatcher.php(421): Illuminate\Events\Dispatcher::Illuminate\Events\{closure}
#20 /vendor/laravel/framework/src/Illuminate/Events/Dispatcher.php(249): Illuminate\Events\Dispatcher::dispatch
#19 /vendor/laravel/framework/src/Illuminate/Queue/Worker.php(618): Illuminate\Queue\Worker::raiseBeforeJobEvent
#18 /vendor/laravel/framework/src/Illuminate/Queue/Worker.php(416): Illuminate\Queue\Worker::process
#17 /vendor/laravel/framework/src/Illuminate/Queue/Worker.php(379): Illuminate\Queue\Worker::runJob
#16 /vendor/laravel/framework/src/Illuminate/Queue/Worker.php(173): Illuminate\Queue\Worker::daemon
#15 /vendor/laravel/framework/src/Illuminate/Queue/Console/WorkCommand.php(150): Illuminate\Queue\Console\WorkCommand::runWorker
#14 /vendor/laravel/framework/src/Illuminate/Queue/Console/WorkCommand.php(134): Illuminate\Queue\Console\WorkCommand::handle
#13 /vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php(36): Illuminate\Container\BoundMethod::Illuminate\Container\{closure}
#12 /vendor/laravel/framework/src/Illuminate/Container/Util.php(41): Illuminate\Container\Util::unwrapIfClosure
#11 /vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php(93): Illuminate\Container\BoundMethod::callBoundMethod
#10 /vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php(37): Illuminate\Container\BoundMethod::call
#9 /vendor/laravel/framework/src/Illuminate/Container/Container.php(651): Illuminate\Container\Container::call
#8 /vendor/laravel/framework/src/Illuminate/Console/Command.php(139): Illuminate\Console\Command::execute
#7 /vendor/symfony/console/Command/Command.php(308): Symfony\Component\Console\Command\Command::run
#6 /vendor/laravel/framework/src/Illuminate/Console/Command.php(124): Illuminate\Console\Command::run
#5 /vendor/symfony/console/Application.php(998): Symfony\Component\Console\Application::doRunCommand
#4 /vendor/symfony/console/Application.php(299): Symfony\Component\Console\Application::doRun
#3 /vendor/symfony/console/Application.php(171): Symfony\Component\Console\Application::run
#2 /vendor/laravel/framework/src/Illuminate/Console/Application.php(102): Illuminate\Console\Application::run
#1 /vendor/laravel/framework/src/Illuminate/Foundation/Console/Kernel.php(129): Illuminate\Foundation\Console\Kernel::handle
#0 /artisan(37): null
```

![image](https://user-images.githubusercontent.com/62543331/208653754-091ac76c-87b1-4bf6-bf6e-93b821aa66d4.png)

From the stack trace it looks like the issue is coming from the package attempting to find the current tenant. The peculiar part is that instead of looking in our landlord database (which in this case would be `application_x_landlord`), it tries to get the tenant from the database of this specific tenant (`'application_x_tenant_2`). This has to imply that the package succesfully fetches the tenant (otherwise - how would it find out the specific tenant database name?), and then tries to find the tenant _again_.

Even stranger is that despite this currently being thrown for every job, the jobs themselves seem to execute just fine.

Does anyone have any insights into why this might be happening? Cheers.

**Comment by slavatarsyurimi** [comment]
Unfortunately still very relevant after upgrading the package to **3.0.0**
Anyone possibly any clue into what is causing tihs? This is flooding our Sentry tracking quite a lot.

**Comment by slavatarsyurimi** [comment]
Hi @freekmurze,

I've been digging into this more lately. My current suspicion is that this happens because of jobs calling other jobs internally using `->dispatchSync()`, making the package think there is no current tenant available at runtime while there actually is one.
Though I haven't yet been able to reproduce it locally with a redis queue, we have been able to overcome the problem on a staging server by applying the following monkeypatch fix to the `MakeQueueTenantAwareAction::listenForJobsBeingProcessed` method:

![image](https://user-images.githubusercontent.com/62543331/216937331-77f616b6-9c6f-4286-bbb2-f2b2df4c20b9.png)
(Same idea for `listenForJobsRetryRequested`)

Is this something you'd like to see a PR for? Logically I can't think of a reason you'd want to find a tenant in a context of a job when you already have a tenant.

**Comment by masterix21** [comment]
Hi @slavatarsyurimi.

IMHO, it's a nonsense fix: why should you make the tenant current if it's already the current one? 

What do I need to understand your solution?

**Comment by slavatarsyurimi** [comment]
Hi @masterix21, sharp. It was supposed to be `=== null` instead of `!== null`.

Unfortunately, upon further investigation this was not the fix after all (monday morning..). I'm gonna look more into it.

**Comment by masterix21** [comment]
Please keep us updated. Thanks.

**Comment by slavatarsyurimi** [comment]
Hi @masterix21,

The issue we had was indeed due to dispatching jobs from another jobs using `dispatchSync()` and has been fixed in [3.0.2](https://github.com/spatie/laravel-multitenancy/compare/3.0.1...3.0.2) by refreshing the current tenant on each job. 

This release however introduced a new issue when using `dispatchSync()` in combination with using database transaction. I'm gonna open a new issue to address it.

**Comment by slavatarsyurimi** [comment]
See https://github.com/spatie/laravel-multitenancy/issues/442
